### PR TITLE
User can enter locked room

### PR DESCRIPTION
**Pull request by user512**
* Implementing: https://xd.adobe.com/view/fd425dbe-5384-44c9-997a-eeee6e886a86-a811/screen/04ee266e-931b-4bde-bcf9-af94c7ac444e/

Screen Shot 2020-09-02 at 9 16 13 PM

### CODE REVIEW

File: convene-web/app/javascript/src/custom_components/button.scss [hunk 1]
```diff
@@ -0,0 +1,9 @@
+.btn {
+  @apply font-bold py-2 px-4 rounded;
+}
+.btn-blue {
+  @apply bg-blue-500 text-white;
+}
+.btn-blue:hover {
+  @apply bg-blue-700;
+}
```
Comment (zspencer) [inline convene-web/app/javascript/src/custom_components/button.scss:9]: Not blocking: I would prefer that we use the same file name as we use for the css classes. I.e. `button.scss` defineds the `.button` class.

**Review by zspencer — COMMENTED** [review]
This looks like 2 or 3 different distinct patches:

1. Lay out the application framing (i.e. heading, plus colors)
2. Review apps have an admin username and password
3. Locked rooms prevent access

And, if I'm feeling spunky, I may include `locked rooms grant access when the right access code is entered` in the 3rd PR; or if I'm too slow; I may split it out into a fourth PR so that we don't have to worry about dealing with merge conflicts if other folks start working on the "pre-video" room landing page.

Overall, this looks really good; and my only recommendation is to take a bit of time to think about how you can split the work so that it can be integrated bit-by-bit.

File: convene-web/app/views/rooms/waiting_room.html.erb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<% room = @current_room %>
+<div class="bg-blue-100 min-h-screen flex items-center justify-between">
+  <div class="m-auto">
+    <div class="bg-blue-500 text-white font-extrabold px-8 py-4 rounded">
+      <%= room.name %>
+      <i class="fas fa-user ml-4"></i>
+    </div>
+  </div>
+    <%= form_tag(workspace_room_path(room.workspace, room), method: :get, class: "m-auto flex flex-col") %>
```
Comment (zspencer) [inline convene-web/app/views/rooms/waiting_room.html.erb:9]: Is this missing a `do`?

File: convene-web/app/views/layouts/video_room.html.erb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  <head>
+    <title><%= "#{current_workspace.name} - #{@current_room.name}" %></title>
+    <meta name="viewport" content="width=device-width, initial-scale=1" />
+    <%= csrf_meta_tags %>
+    <%= csp_meta_tag %>
+
+    <%= stylesheet_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
+    <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
+    <% if current_workspace&.jitsi_meet_domain %>
+      <script src='https://<%= current_workspace.jitsi_meet_domain %>/external_api.js'></script>
```
[the remark this answers is not in the document]
Comment (user512) [reply convene-web/app/views/layouts/video_room.html.erb:12]: I have a few instance where the video room won't start after switching to async.
Maybe I should consider using defer where the script will run right before the document is ready?

[the remark this answers is not in the document]
Comment (zspencer) [reply convene-web/app/views/layouts/video_room.html.erb]: Ahh, that makes sense.

File: convene-web/app/views/layouts/_header.html.erb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+  <header class="header">
+    <div class="items-container">
+      <h1><%= current_workspace.client.name %></h1>
+      <div>
+        <h3><%= current_workspace.name %></h3>
+        <h3 class="font-hairline">video meeting spaces</h3>
+      </div>
+    </div>
```
Comment (zspencer) [inline convene-web/app/views/layouts/_header.html.erb:8]: Not blocking, but my understanding of the header layout @colombene put together is that the left hand side is the current workspace and the right hand side is the Convene branding. This appears to invert that.

**Review by zspencer — APPROVED** [review]
This is a great paper-wall! There are a few things that I think we’ll want to improve on moving forward:

1.It feels to me like it can be hard to merge changes when they cover so much. How would you feel about pushing a bit harder on making really small-but-useful changes? I.e. if you notice that we need a layout change, make a new branch to make the layout change and push that up?
2. It looks like you undid changes to the feature tests that @cjoulain and I just implemented. It may be worth doing a `git log` to get context if you notice something doesn’t seem quite the way you would want it to be; and it’s also changed recently.

File: features/locking-rooms.feature [hunk 5]
```diff
@@ -19,17 +19,25 @@ Feature: Locking Rooms
...
-    And a Workspace Member may not enter the Room after providing the wrong Room Key
-    And a Guest may enter the Room after providing the correct Room Key
-    And a Guest may not enter the Room after providing the wrong Room Key
+  Scenario: Workspace Admin entering a Locked Room
     # We're not sure if this is actually a good idea, we should check with
     # design / product to make sure that it makes sense for admins to be able
     # to barge in like a grumpy parent on prom night.
-    And a Workspace Admin may enter the Room without providing a Room Key
+    Given the Workspace Admin is on the "System Test" Workspace Dashboard
+    And a Workspace Admin may enter the "Listed Locked Room 1" without providing a Room Key
```
[the remark this answers is not in the document]
Comment (user512) [reply features/locking-rooms.feature:40]: I didn't know it's grouped this way intentionally, I can revert it back.
I was trying to reuse test steps to save time on writing feature test.

File: features/steps/room_steps.js [hunk 6]
```diff
@@ -45,14 +45,15 @@ Then('a {actor} may enter the Room without providing {roomKey}', function (actor
   return 'pending';
 });
 
-Then('a {actor} may not enter the {room} after providing {roomKey}', function (actor, room, roomKey) {
-  // Write code here that turns the phrase above into concrete actions
+Then('a {actor} may not enter the {room} after providing {roomKey}', async function (actor, room, roomKey) {
+  // await this.workspace.enterRoomWithAccessCode(room, 'wrongAccessCode');
```
[the remark this answers is not in the document]
Comment (user512) [reply features/steps/room_steps.js:49]: Yes, I was planning to add assertions here once I completed the error state on invalid access code so there's some interesting thing to test.

File: convene-web/app/views/layouts/video_room.html.erb [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
+    <% if current_workspace&.jitsi_meet_domain %>
+      <script defer src='https://<%= current_workspace.jitsi_meet_domain %>/external_api.js'></script>
+    <% end %>
+  </head>
+
+  <body>
+    <%= yield %>
+  </body>
+</html>
```
[the remark this answers is not in the document]
Comment (user512) [reply convene-web/app/views/layouts/video_room.html.erb:19]: I think the design layout in video room and workspace is quite different so instinctively this feels right.
I think it'd makes more sense once I have the video room bottom bar built out.

[the remark this answers is not in the document]
Comment (zspencer) [reply convene-web/app/views/layouts/video_room.html.erb:19]: Gotcha. 

I think we may want to explore ways to make the design work without multiple layouts. We can leverage things like [`content_for`](https://guides.rubyonrails.org/layouts_and_rendering.html#using-the-content-for-method) and partials. And, (if necessary) we can use [nested layouts](https://guides.rubyonrails.org/layouts_and_rendering.html#using-nested-layouts).

**Pull request merged**